src/v/kafka/server/handlers/produce.cc

```diff
@@ -267,10 +267,41 @@ static partition_produce_stages produce_topic_partition(
           .partition_index = ntp.tp.partition,
           .error_code = error_code::unknown_topic_or_partition});
     }
+
+    auto const broker_time = model::timestamp::now();
```

Comment:
nit: Can we pull out the timestamp handling into it's own function? I think that will aid readability.

Basically we need something like: `std::optional<model::timestamp> batch_timestamp(const record_batch_header&)`

Inside this function we can do all the validation and alerting, and if a timestamp is returned then it overwrites the batch header timestamp.

EDIT: I suppose you really need a timestamp_type too.

Reply:
i think i get what you are saying. Maybe a validate_batch_timestamp? I would need to bring some links too, for the probe

Reply:
`validate_batch_timestamp` is a better name yes :)

Ah right RE probe and additional parameters, I was a little naive, but mostly pulling this out I think would be nice.